### PR TITLE
Blocking calls force their replies evaluation

### PR DESCRIPTION
**Pull request by qrilka**
Blocking commands could easily lead to reace conditions
between multiple `runRedis` calls as it is demonstraded in #23.
So to prevent this from happenning those commands are set
to force their replies evaluation up to WHNF

### CODE REVIEW

**Comment by informatikr** [comment]
Hi Kiril,

I think your approach is certainly an option, since the number of blocking commands is relatively small. What I don't like about it, is that it is one more thing to keep in mind when maintaining hedis. For example when a new command is added to Redis, you can't "just add it" to hedis, but also have to check if it blocks. This is also code that is hard to test in the automatic test suite.

The problem found in #23 is that a blocked connection is returned to the resource pool. This connection is then given to other `runRedis` calls, but might never unblock again. Especially if the unblocking command is sent over the same connection and we deadlock.

So another approach to fix this would be that `Redis` keeps track of the last `Reply` is "received". Not the evaluated Reply, but the thunk. Basically the `r` in Database.Redis.Core.sendRequest. It would then be sufficient if `runRedis` evaluated the last thunk before returning.

This would break the "pipelining between threads", but that was never planned design and isn't an important feature in my opinion.

To summarize:
* Keep track of the most recently received reply-thunk in Redis (e.g. make Redis a `MonadState Reply`)
* `evaluate` the most recent reply before returning the connection to the pool in `runRedis`.

Thank you for your effort, and I hope that helps.

**Comment by qrilka** [comment]
I was thinking about that option too but it seems to me that it will break the promise in the current hedis haddocks to use automatic pipelining across several calls to `runRedis`. I.e. not between threads but even for the same thread. And that appears to me to be a planned design, isn't it?
Of course design (and haddocks) could be changed but could such change have some negative impact on hedis users?

**Comment by informatikr** [comment]
I never planned it. It was just a consequence of the implementation and I put it in the documentation. I don't know if anybody makes use of it.

In my opinion this "hypothetical feature" should be removed if it means that a tangible benefit can be gained elsewhere.

**Comment by qrilka** [comment]
Ok, great, then I'll create another PR with that alternative implementation,
Thanks a lot.

**Comment by qrilka** [comment]
And #45 is a follow-up for this one :)
